### DISCUSSION TITLE
Unit testing Akita effects

### DISCUSSION THREAD

**Discussion by csaszika** [salesforce/akita#633] [Q&A]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[?] Bug report  
[ ] Performance issue
[?] Feature request
[X] Documentation issue or request
[ ] Support request
[ ] Other... Please describe:
</code></pre>

## Current behavior

I found no examples.

## Expected behavior

I would like to unit test my effects.

### Working code
<pre><code>
import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@datorama/akita-ng-effects';
import { of } from 'rxjs';
import { catchError, exhaustMap, map } from 'rxjs/operators';

import { CalendarSchema } from '../../models/calendar-schema';
import { CalendarSchemaService } from '../../services/calendar-schema/calendar-schema.service';
import { saveCalendarSchema, saveCalendarSchemaFailed, saveCalendarSchemaSuccess } from './calendar-schema.actions';

@Injectable({
    providedIn: 'root',
})
export class CalendarSchemaEffects {
    constructor(private readonly _actions$: Actions, private readonly _calendarSchemaService: CalendarSchemaService) {}

    saveCalendarSchema$ = createEffect(
        () =>
            this._actions$.pipe(
                ofType(saveCalendarSchema),
                exhaustMap((data: { schemas: CalendarSchema[] }) =>
                    this._calendarSchemaService.saveCalendarSchemas$(data.schemas).pipe(
                        map((schemas: CalendarSchema[]) => saveCalendarSchemaSuccess({ schemas })),
                        catchError((schemas: CalendarSchema[]) => of(saveCalendarSchemaSuccess({ schemas }))),
                    ),
                ),
            ),
        { dispatch: true },
    );

    saveCalendarSchemaSuccess$ = createEffect(() =>
        this._actions$.pipe(
            ofType(saveCalendarSchemaSuccess),
            map((data: { schemas: CalendarSchema[] }) => this._calendarSchemaService.addCalendarSchemas(data.schemas)),
        ),
    );

    saveCalendarSchemaFailed$ = createEffect(
        () =>
            this._actions$.pipe(
                ofType(saveCalendarSchemaFailed),
                map((data: { schemas: CalendarSchema[] }) => console.log(data)),
            ),
        { dispatch: false },
    );
}

</code></pre>

### Test
<pre><code>
import { TestBed } from '@angular/core/testing';
import { Actions, AkitaNgEffectsModule } from '@datorama/akita-ng-effects';
import { SpyObject, mockProvider } from '@ngneat/spectator';
import { cold } from 'jasmine-marbles';

import { calendarSchemaFactory } from '../../models/calendar-schema.factory';
import { CalendarSchemaService } from '../../services/calendar-schema/calendar-schema.service';
import { saveCalendarSchema, saveCalendarSchemaSuccess } from './calendar-schema.actions';
import { CalendarSchemaEffects } from './calendar-schema.effects';

const SCHEMA_1 = calendarSchemaFactory.withSchemaName('schema').withWeekDays([1, 5]).build();
const SCHEMA_2 = calendarSchemaFactory.withSchemaName('schema2').withWeekDays([1, 5]).build();

describe('CalendarSchemaEffects', () => {
    let effects: CalendarSchemaEffects;
    let calendarSchemaService: SpyObject<CalendarSchemaService>;
    let actions$: Actions;

    const schemas = { schemas: [SCHEMA_1, SCHEMA_2] };

    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [AkitaNgEffectsModule.forRoot([CalendarSchemaEffects])], // <-- THIS DOES NOT HELP
            providers: [
                CalendarSchemaEffects,
                { provide: CalendarSchemaService, useValue: mockProvider(CalendarSchemaService).useFactory() as SpyObject<CalendarSchemaService>},
                Actions,
            ],
        });
        effects = TestBed.inject(CalendarSchemaEffects);
        actions$ = TestBed.inject(Actions);
        calendarSchemaService = TestBed.inject(CalendarSchemaService) as SpyObject<CalendarSchemaService>;
    });

    it('should call the mock backend api and return the results', () => {
        const action = saveCalendarSchema(schemas);
        const result = saveCalendarSchemaSuccess(schemas);

        const response = cold('-a|', { a: [SCHEMA_1, SCHEMA_2] });
        const expected = cold('-b|', { b: result });
        calendarSchemaService.saveCalendarSchemas$.and.returnValue(response);

        actions$.dispatch(action);

        expect(effects.saveCalendarSchema$).toBeObservable(expected);
    });
});
</code></pre>

This is what have I tried but it does not work, my effects is not called and I got this error message:
<pre><code>
Error: 
	    Expected: -b|,
	    Received: ,
	    
	    Expected:
	    [{"frame":10,"notification":{"kind":"N","value":{"schemas":[{"id":0,"itemNumbers":[],"schemaName":"schema","weekDays":[1,5]},{"id":0,"itemNumbers":[],"schemaName":"schema","weekDays":[1,5]}],"type":"[Calendar schema] Save calendar schema success"},"hasValue":true}},{"frame":20,"notification":{"kind":"C","hasValue":false}}]
	    
	    Received:
	    [],
	  
	    at <Jasmine>
	    at UserContext.<anonymous> (src/app/meat-order/components/calendar-schema-creator/calendar-schema.effects.spec.ts:48:45)
	    at ZoneDelegate.invoke (node_modules/zone.js/dist/zone-evergreen.js:368:1)
	    at ProxyZoneSpec.onInvoke (node_modules/zone.js/dist/zone-testing.js:300:1)
</code></pre>

## Environment

<pre><code>
Angular version: 11.0.4
 
For Tooling issues:
- Node version: 15.2.0

Thanks for help in advance.

**Comment by juri-sinitson** [comment]
Hi,

I just pull-requested simple examples which I tested before.
See https://github.com/datorama/akita/pull/622

Hope they'll be helpful for you. 

I think an `HttpTestingController` and expectations in `subscribe`-blocks may be a an alternative to testing with marbles. If you want to stay by marbles I'm missing the call of change detection and flushing with a test scheduler in your example.

See also: https://angular.io/guide/testing-components-scenarios#component-marble-tests

**Comment by csaszika** [comment]
Hi,

thanks for the examples. Yes I realized unfortunately marbles are not work really with akita effects. I tried it because it was a clean testing method with ngrx effects. (In akita effects there are some missing properties in test like `marble`.)

Yes I did not use change detection and test scheduler because I wanted to unit test only the effect.
Anyway I checked your examples @juri-sinitson and those were good but not really a unit tests what I was looking for.

After some investigation I found a way which is comfortable and clean for me and perhaps useful for the others to unit test akita effects.

Full test:
<pre><code>
import { TestBed } from '@angular/core/testing';
import { Actions } from '@datorama/akita-ng-effects';
import { SpyObject, mockProvider } from '@ngneat/spectator';
import { of, throwError } from 'rxjs';

import { calendarSchemaFactory } from '../../models/calendar-schema.factory';
import { CalendarSchemaService } from '../../services/calendar-schema/calendar-schema.service';
import { saveCalendarSchema, saveCalendarSchemaFailed, saveCalendarSchemaSuccess } from './calendar-schema.actions';
import { CalendarSchemaEffects } from './calendar-schema.effects';

const SCHEMA_1 = calendarSchemaFactory.withSchemaName('schema').withWeekDays([1, 5]).build();
const SCHEMA_2 = calendarSchemaFactory.withSchemaName('schema').withWeekDays([1, 5]).build();

describe('CalendarSchemaEffects', () => {
    let effects: CalendarSchemaEffects;
    let calendarSchemaService: SpyObject<CalendarSchemaService>;
    let actions$: Actions;

    const schemasPayload = { schemas: [SCHEMA_1, SCHEMA_2] };

    beforeEach(() => {
        TestBed.configureTestingModule({
            providers: [
                CalendarSchemaEffects,
                {
                    provide: CalendarSchemaService,
                    useValue: mockProvider(CalendarSchemaService).useFactory() as SpyObject<CalendarSchemaService>,
                },
            ],
        }).compileComponents();
        effects = TestBed.inject(CalendarSchemaEffects);
        actions$ = TestBed.inject(Actions);
        calendarSchemaService = TestBed.inject(CalendarSchemaService) as SpyObject<CalendarSchemaService>;
    });

    describe('#saveCalendarSchema$', () => {
        it('should return success action', (done) => {
            const action = saveCalendarSchema(schemasPayload);
            const result = saveCalendarSchemaSuccess(schemasPayload);

            calendarSchemaService.saveCalendarSchemas$.and.returnValue(of([SCHEMA_1, SCHEMA_2]));

            effects.saveCalendarSchema$.subscribe((action) => {
                expect(action).toEqual(result);
                done();
            });

            actions$.dispatch(action);
        });

        it('should return failed action', (done) => {
            const error = 'FAIL';
            const action = saveCalendarSchema(schemasPayload);
            const result = saveCalendarSchemaFailed({ error });

            calendarSchemaService.saveCalendarSchemas$.and.returnValue(throwError(error));

            effects.saveCalendarSchema$.subscribe((action) => {
                expect(action).toEqual(result);
                done();
            });

            actions$.dispatch(action);
        });
    });

    describe('#saveCalendarSchemaSuccess$', () => {
        it('should update store via service', (done) => {
            const action = saveCalendarSchemaSuccess(schemasPayload);

            effects.saveCalendarSchemaSuccess$.subscribe(() => {
                expect(calendarSchemaService.addCalendarSchemas).toHaveBeenCalledTimes(1);
                expect(calendarSchemaService.addCalendarSchemas).toHaveBeenCalledWith(schemasPayload.schemas);
                done();
            });

            actions$.dispatch(action);
        });
    });

    describe('#saveCalendarSchemaFailed$', () => {
        it('should update store via service', (done) => {
            const action = saveCalendarSchemaFailed({ error: 'fail' });

            spyOn(console, 'log').and.callThrough();

            effects.saveCalendarSchemaFailed$.subscribe(() => {
                expect(console.log).toHaveBeenCalledTimes(1);
                expect(console.log).toHaveBeenCalledWith(action);
                done();
            });

            actions$.dispatch(action);
        });
    });
});
</code></pre>

So as I found a solution, this issue can be closed.

**Comment by juri-sinitson** [comment]
Hi,

when I understand right, the aim of your issue was not only to 1. find a solution, but also 2. to improve the documentation with a simple example or at least with a tipp to subscribe to an effect. In my opinion the 1 is completed but 2 not.

So please don't close this issue, until the point 2 is completed.

I think the aim is, that other developers will not have to do the same work again and again you and I did.

**Comment by yackinn** [comment]
I'll inlcude an example for testing in the docs. 
Marble testing isnÄt supported at this point. We're not planning on doing so either. Testing without marbles should suffice for now. Feature enhancements are prioritized.
